app/assets/javascripts/discourse/controllers/reorder-categories.js.es6

```diff
@@ -31,123 +26,113 @@ export default Controller.extend(ModalFunctionality, Ember.Evented, {
-              desiredIdx = cats.indexOf(tmpCat);
-              otherCat = tmpCat;
-              break;
-            }
-          }
-        } else {
-          // We want to move up
-          cats.forEach(function(tmpCat) {
-            if (tmpCat.get("id") === parentIdx) {
-              desiredIdx = cats.indexOf(tmpCat);
-              otherCat = tmpCat;
-            }
-          });
-        }
-      }
+  move(category, direction) {
+    const currentIdx = category.get("position");
+    const targetIdx = currentIdx + direction;
 
-      otherCat.set("position", curIdx);
-      cat.set("position", desiredIdx);
-      this.send("commit");
+    if (targetIdx < 0 || targetIdx >= this.categoriesOrdered.length) {
+      return;
     }
-  },
-
-  /**
-    1. Make sure all categories have unique position numbers.
-    2. Place sub-categories after their parent categories while maintaining the
-        same relative order.
-
-        e.g.
-          parent/c1         parent
-          parent      =>    parent/c1
-          other             parent/c2
-          parent/c2         other
-  **/
-  fixIndices() {
-    const categories = this.categoriesOrdered;
-    const subcategories = {};
-
-    categories.forEach(category => {
-      const parentCategoryId = category.get("parent_category_id");
-
-      if (parentCategoryId) {
-        subcategories[parentCategoryId] = subcategories[parentCategoryId] || [];
-        subcategories[parentCategoryId].push(category);
-      }
-    });
 
-    for (let i = 0, position = 0; i < categories.get("length"); ++i) {
-      const category = categories.objectAt(i);
+    let otherCategory = this.categoriesOrdered.find(
+      c =>
+        category.get("parent_category_id") === c.get("parent_category_id") &&
+        c.get("position") === targetIdx
+    );
+
+    // Special case to handle "move to top" or "move to bottom"
+    // `otherCategory` should be the first category if `direction < 0` or the
+    // last category if `direction > 0`.
+    if (!otherCategory) {
+      this.categoriesOrdered.forEach(c => {
+        if (
+          category.get("parent_category_id") === c.get("parent_category_id") &&
+          (!otherCategory || direction > 0)
+        ) {
+          otherCategory = c;
+        }
+      });
+    }
 
-      if (!category.get("parent_category_id")) {
-        category.set("position", position++);
-        (subcategories[category.get("id")] || []).forEach(subcategory =>
-          subcategory.set("position", position++)
-        );
-      }
+    if (category.get("id") === otherCategory.get("id")) {
+      return;
     }
+
+    category.set("position", targetIdx);
+    otherCategory.set("position", currentIdx);
+
+    this.reorder();
   },
 
   actions: {
-    change(cat, e) {
-      let position = parseInt($(e.target).val(), 10);
-      let amount = Math.min(
-        Math.max(position, 0),
+    change(category, event) {
+      let newPosition = parseInt($(event.target).val(), 10);
```

Comment:
Is there a way we could do this without using jQuery?

Reply:
Looks like a `DOMElement` is already passed. I do not know the reason why jQuery was used, probably just to make the test work.